### PR TITLE
chore: development documentation

### PR DESCRIPTION
**Pull request by MilGard91**
## Context

Include relevant motivation and context. A link to the issue is enough if it contains all the relevant 
information.

## Changes proposed in this pull request

Include a summary of the change.

## Test plan

Please describe the tests cases that you ran to verify your changes. Add further instructions on 
how to run them if needed (i.e. migration / deployment scripts, env vars, etc).

## Checklist

- [ ] Rebased the base branch
- [ ] Attached corresponding Github issue
- [ ] Prefixed the name with the type of change (i.e. feat, chore, test)
- [ ] Performed a self-review of my own code
- [ ] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [ ] Didn't generate new warnings
- [ ] Didn't generate failures on existing tests
- [ ] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: docs/Ramm.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  uint nxmB,
+  uint budget
+);
+```
+
+| Return         | Description                         |
+|----------------|-------------------------------------|
+| `ethReserve`   | ETH reserve in the contract         |
+| `nxmA`         | NXM buy price                       |
+| `nxmB`         | NXM sell price                      |
```
[the remark this answers is not in the document]
Comment (shark0der) [reply docs/Ramm.md:129]: Ah and `ethReserve` is the available eth liquidity in the virtual pools

File: docs/Ramm.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+| `budget`       | Available ETH budget for liquidity  |
+
+### `getSpotPrices`
+Returns the current spot prices for NXM buy and sell operations.
```
Comment (shark0der) [inline docs/Ramm.md:133]: Worth mentioning that when swapping the prices are on a uniswap v2 -like curve and the spot prices are just where the price curve starts so someone wouldn't expect to multiply this by the amount and expect that result.

File: docs/Ramm.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+function getSpotPrices() external view returns (uint spotPriceA, uint spotPriceB);
+```
+
+| Return         | Description            |
+|----------------|------------------------|
+| `spotPriceA`   | Current NXM buy price  |
+| `spotPriceB`   | Current NXM sell price |
+
+### `getBookValue`
+Returns the current book value of NXM.
```
Comment (shark0der) [inline docs/Ramm.md:145]: Worth adding a tiny explanation of what the book value is

File: docs/Ramm.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+uint public constant TARGET_LIQUIDITY = 5_000 ether;
+uint public constant FAST_LIQUIDITY_SPEED = 1_500 ether;
+uint public constant NORMAL_RATCHET_SPEED = 400;
+uint public constant FAST_RATCHET_SPEED = 5_000;
+```
+
+## Mutative Functions
+
+### `swap`
+Allows users to swap NXM tokens for ETH or vice versa.
```
Comment (shark0der) [inline docs/Ramm.md:62]: This function can and probably should be used from offchain to estimate the amount of nxm or eth one would if they swap by using `ramm.callStatic.swap(nxmIn, 0, future_timestamp)` (`eth_call`). They should add the slippage when swapping.

File: docs/Ramm.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
+
+# Ramm Contract Developer Documentation
+
+## Overview
+
+The `Ramm` contract is designed to manage liquidity and facilitate swaps between NXM tokens and ETH. It provides a dynamic pricing mechanism through ratcheting and circuit breakers to maintain stability. The contract includes features for managing ETH and NXM reserves, along with ensuring that swaps operate within predefined parameters and limits.
```
Comment (shark0der) [inline docs/Ramm.md:6]: The `Ramm` contract is designed to allow swaps between NXM tokens and ETH. Internally it works by simulating 2 uniswap v2 -like pools which have their liquidity adjusted using liquidity injection and a ratcheting mechanism.

File: docs/StakingPool.md [hunk 6]
```diff
@@ -0,0 +1,289 @@
+
+# StakingPool Contract Developer Documentation
+
+## Overview
+
+The `StakingPool` contract is a smart contract that manages staking and rewards distribution for a staking pool in a decentralized cover protocol. It interacts with multiple other contracts to provide staking functionalities, including staking NFT issuance, stake deposit and withdrawal, and rewards allocation. The contract ensures that stakes are managed over defined periods (tranches) and calculates rewards based on the stake's contribution to the pool.
```
Comment (shark0der) [inline docs/StakingPool.md:6]: The `StakingPool` contract manages nxm staking and capacity allocations for purchased covers. When depositing an NFT is minted which is used to track stake ownership.

File: docs/StakingPool.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+### Buckets
+Groupings used to manage rewards over time.
+
+```solidity
+uint public constant BUCKET_DURATION = 28 days;
+```
+
+```solidity
+uint public constant BUCKET_DURATION = 28 days;
+```
```
Comment (shark0der) [inline docs/StakingPool.md:43]: dupe

File: docs/StakingPool.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+uint public constant TRANCHE_DURATION = 91 days;
+uint public constant MAX_ACTIVE_TRANCHES = 8; // 7 whole quarters + 1 partial quarter
+```
+
+Formula for current tranche id:
+```solidity
+uint currentTrancheId = block.timestamp / TRANCHE_DURATION;
+```
+### Buckets
+Groupings used to manage rewards over time.
```
Comment (shark0der) [inline docs/StakingPool.md:35]: And cover allocation expirations

File: docs/StakingPool.md [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+```solidity
+uint public constant BUCKET_DURATION = 28 days;
+```
+Formula for current bucket id:
+```solidity
+    uint currentBucketId = block.timestamp / BUCKET_DURATION;
+```
+
+### Allocations
+Groupings of used capacity per tranche.
```
Comment (shark0der) [inline docs/StakingPool.md:50]: Amount of used capacity by individual covers

File: docs/StakingPool.md [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+## Mutative Functions
+
+### `depositTo`
+Allows users to deposit NXM into the pool, receiving stake and rewards shares in return. Supports deposits to specific tranches and allows adding to an existing deposit.
```
Comment (shark0der) [inline docs/StakingPool.md:55]: s/receiving/creating/

File: docs/StakingPool.md [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+```
+| Parameter        | Description                                                                                                 |
+|------------------|-------------------------------------------------------------------------------------------------------------|
+| `amount`         | The amount to deposit.                                                                                      |
+| `trancheId`      | The ID of the tranche to deposit into.                                                                      |
+| `requestTokenId` | The ID of the request token (0 for a new deposit, or use an existing token ID to add to a previous deposit. |
+| `destination`    | The address to send the Staking NFT token to.                                                               |
+
+### `withdraw`
+Allows users to withdraw their stake and/or rewards from specific tranches, only from expired tranches.
```
Comment (shark0der) [inline docs/StakingPool.md:68]: "only from expired tranches" - only applies to stake, rewards can be withdrawn at any time

File: docs/StakingPool.md [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+```solidity
+function withdraw(uint tokenId, bool withdrawStake, bool withdrawRewards, uint[] memory trancheIds) external;
+```
+
+| Parameter         | Description                                       |
+|-------------------|---------------------------------------------------|
+| `tokenId`         | The ID of the staking NFT token.                  |
+| `withdrawStake`   | Whether to withdraw the stake.                    |
+| `withdrawRewards` | Whether to withdraw the rewards.                  |
+| `trancheIds`      | The IDs of the expired tranches to withdraw from. |
```
Comment (shark0der) [inline docs/StakingPool.md:80]: d/expired/

File: docs/StakingPool.md [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+
+### `extendDeposit`
+Extends the duration of a deposit by moving it from an active tranche to a future tranche.
```
Comment (shark0der) [inline docs/StakingPool.md:83]: "from an active tranche" -> false. tranche can be active or expired, doesn't matter

File: docs/StakingPool.md [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+  uint stakeSharesSupplyAtExpiry
+);
+```
+
+| Parameter        | Description            |
+|------------------|------------------------|
+| `trancheId`      | The ID of the tranche. |
+
+### `getActiveAllocations`
+Returns an array of the active allocations for a product per tranche.
```
Comment (shark0der) [inline docs/StakingPool.md:148]: s/per tranche/for currently active tranches/

File: docs/StakingPool.md [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+```solidity
+function getActiveAllocations(uint productId)  public view returns (uint[] memory trancheAllocations);
+```
+
+| Parameter          | Description                          |
+|--------------------|--------------------------------------|
+| `productId`        | The ID of the product.               |
+
+### `getActiveTrancheCapacities`
+Returns an array of the active tranche capacities and total capacity for a product.
```
Comment (shark0der) [inline docs/StakingPool.md:159]: Returns an array with the total capacities for the currently active tranches for a product.

**Comment by rackstar** [comment]
closing in favour of https://github.com/NexusMutual/smart-contracts/pull/1295

**Pull request closed without merging**